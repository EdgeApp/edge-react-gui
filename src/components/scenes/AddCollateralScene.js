// @flow

import s from '../../locales/strings.js'
import { type NavigationProp, type RouteProp } from '../../types/routerTypes.js'
import { ManageCollateralScene } from './ManageCollateralScene.js'

type Props = {
  navigation: NavigationProp<'addCollateralScene'>,
  route: RouteProp<'addCollateralScene'>
}

export const AddCollateralScene = (props: Props) => {
  const { navigation, route } = props
  const { borrowEngine } = route.params

  return ManageCollateralScene({
    borrowEngine,
    defaultTokenId: borrowEngine.collaterals[0].tokenId,
    action: async req => await borrowEngine.deposit(req),
    actionWallet: 'fromWallet',
    ltvType: 'collaterals',
    ltvChange: 'decrease',

    showExchangeRateTile: true,
    showTotalDebtTile: true,
    showTotalCollateralTile: true,

    headerText: s.strings.loan_add_collateral,
    goBack: () => navigation.pop()
  })
}
