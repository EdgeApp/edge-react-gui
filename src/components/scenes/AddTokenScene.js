// @flow

import * as React from 'react'
import { ActivityIndicator, Alert, ScrollView, StyleSheet, View } from 'react-native'

import { addNewToken } from '../../actions/AddTokenActions.js'
import { MAX_TOKEN_CODE_CHARACTERS } from '../../constants/WalletAndCurrencyConstants.js'
import s from '../../locales/strings.js'
import { PrimaryButton } from '../../modules/UI/components/Buttons/PrimaryButton.ui.js'
import { FormattedText as Text } from '../../modules/UI/components/FormattedText/FormattedText.ui.js'
import { THEME } from '../../theme/variables/airbitz.js'
import { connect } from '../../types/reactRedux.js'
import { type RouteProp } from '../../types/routerTypes.js'
import type { CustomTokenInfo, GuiWallet } from '../../types/types.js'
import { scale } from '../../util/scaling.js'
import { decimalPlacesToDenomination } from '../../util/utils.js'
import { FormField } from '../common/FormField.js'
import { SceneWrapper } from '../common/SceneWrapper.js'

type OwnProps = {
  route: RouteProp<'addToken'>
}

type DispatchProps = {
  addNewToken: (walletId: string, currencyName: string, currencyCode: string, contractAddress: string, denomination: string, type: string) => void
}

type StateProps = {
  addTokenPending: boolean,
  currentCustomTokens: CustomTokenInfo[],
  wallet: GuiWallet
}

type State = {
  currencyName: string,
  currencyCode: string,
  contractAddress: string,
  decimalPlaces: string,
  enabled?: boolean
}

type Props = OwnProps & StateProps & DispatchProps

class AddToken extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    const { route } = this.props
    const { currencyName = '', currencyCode = '', contractAddress = '', decimalPlaces = '' } = route.params
    this.state = {
      currencyName,
      currencyCode,
      contractAddress,
      decimalPlaces
    }
  }

  render() {
    const { addTokenPending } = this.props
    return (
      <SceneWrapper background="body">
        <ScrollView style={styles.container}>
          <View style={styles.instructionalArea}>
            <Text style={styles.instructionalText}>{s.strings.addtoken_top_instructions}</Text>
          </View>
          <View style={styles.nameArea}>
            <FormField
              value={this.state.currencyName}
              onChangeText={this.onChangeName}
              autoCapitalize="words"
              autoFocus
              label={s.strings.addtoken_name_input_text}
              returnKeyType="done"
              autoCorrect={false}
            />
          </View>
          <View style={styles.currencyCodeArea}>
            <FormField
              value={this.state.currencyCode}
              onChangeText={this.onChangeCurrencyCode}
              autoCapitalize="characters"
              label={s.strings.addtoken_currency_code_input_text}
              returnKeyType="done"
              autoCorrect={false}
              maxLength={MAX_TOKEN_CODE_CHARACTERS}
            />
          </View>
          <View style={styles.contractAddressArea}>
            <FormField
              value={this.state.contractAddress}
              onChangeText={this.onChangeContractAddress}
              label={s.strings.addtoken_contract_address_input_text}
              returnKeyType="done"
              autoCorrect={false}
            />
          </View>
          <View style={styles.decimalPlacesArea}>
            <FormField
              value={this.state.decimalPlaces}
              onChangeText={this.onChangeDecimalPlaces}
              label={s.strings.addtoken_denomination_input_text}
              returnKeyType="done"
              autoCorrect={false}
              keyboardType="numeric"
            />
          </View>
          <View style={styles.buttonsArea}>
            <PrimaryButton style={styles.saveButton} onPress={this._onSave}>
              {addTokenPending ? <ActivityIndicator color={THEME.COLORS.ACCENT_MINT} /> : <PrimaryButton.Text>{s.strings.string_save}</PrimaryButton.Text>}
            </PrimaryButton>
          </View>
          <View style={styles.bottomPaddingForKeyboard} />
        </ScrollView>
      </SceneWrapper>
    )
  }

  onChangeName = (input: string) => {
    this.setState({
      currencyName: input
    })
  }

  onChangeCurrencyCode = (input: string) => {
    this.setState({
      currencyCode: input
    })
  }

  onChangeDecimalPlaces = (input: string) => {
    this.setState({
      decimalPlaces: input
    })
  }

  onChangeContractAddress = (input: string) => {
    this.setState({
      contractAddress: input.trim()
    })
  }

  _onSave = () => {
    const currencyCode = this.state.currencyCode.toUpperCase()
    this.setState(
      {
        currencyCode
      },
      () => {
        const { route, currentCustomTokens, wallet, addNewToken } = this.props
        const { currencyName, decimalPlaces, contractAddress } = this.state
        const { walletId } = route.params

        const currentCustomTokenIndex = currentCustomTokens.findIndex(item => item.currencyCode === currencyCode)
        const metaTokensIndex = wallet.metaTokens.findIndex(item => item.currencyCode === currencyCode)
        // if token is hard-coded into wallets of this type
        if (metaTokensIndex >= 0) Alert.alert(s.strings.manage_tokens_duplicate_currency_code)
        // if that token already exists and is visible (ie not deleted)
        if (currentCustomTokenIndex >= 0 && currentCustomTokens[currentCustomTokenIndex].isVisible !== false) {
          Alert.alert(s.strings.manage_tokens_duplicate_currency_code)
        } else {
          if (currencyName && currencyCode && decimalPlaces && contractAddress) {
            const denomination = decimalPlacesToDenomination(decimalPlaces)
            addNewToken(walletId, currencyName, currencyCode, contractAddress, denomination, wallet.type)
          } else {
            Alert.alert(s.strings.addtoken_invalid_information)
          }
        }
      }
    )
  }
}

const rawStyles = {
  container: {
    flex: 1,
    paddingHorizontal: scale(20)
  },

  instructionalArea: {
    paddingVertical: scale(16),
    paddingHorizontal: scale(20)
  },
  instructionalText: {
    fontSize: scale(16),
    textAlign: 'center'
  },

  nameArea: {
    height: scale(70)
  },
  currencyCodeArea: {
    height: scale(70)
  },
  contractAddressArea: {
    height: scale(70)
  },
  decimalPlacesArea: {
    height: scale(70)
  },
  buttonsArea: {
    marginTop: scale(16),
    height: scale(52),
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    alignSelf: 'flex-end',
    paddingVertical: scale(4)
  },
  saveButton: {
    flex: 1,
    marginLeft: scale(2),
    backgroundColor: THEME.COLORS.SECONDARY,
    borderRadius: 3
  },
  bottomPaddingForKeyboard: {
    height: scale(300)
  }
}
const styles: typeof rawStyles = StyleSheet.create(rawStyles)

export const AddTokenScene = connect<StateProps, DispatchProps, OwnProps>(
  (state, { route: { params } }) => ({
    addTokenPending: state.ui.wallets.addTokenPending,
    currentCustomTokens: state.ui.settings.customTokens,
    wallet: state.ui.wallets.byId[params.walletId]
  }),
  dispatch => ({
    addNewToken(walletId: string, currencyName: string, currencyCode: string, contractAddress: string, denomination: string, walletType: string) {
      dispatch(addNewToken(walletId, currencyName, currencyCode, contractAddress, denomination, walletType))
    }
  })
)(AddToken)
