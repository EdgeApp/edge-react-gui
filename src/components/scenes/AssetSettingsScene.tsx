import * as React from 'react'
import FontAwesomeIcon from 'react-native-vector-icons/FontAwesome'

import { CURRENCY_SETTINGS_KEYS } from '../../constants/WalletAndCurrencyConstants'
import { useHandler } from '../../hooks/useHandler'
import { lstrings } from '../../locales/strings'
import { useSelector } from '../../types/reactRedux'
import { EdgeAppSceneProps } from '../../types/routerTypes'
import { SceneWrapper } from '../common/SceneWrapper'
import { CryptoIcon } from '../icons/CryptoIcon'
import { showToast } from '../services/AirshipInstance'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { SettingsTappableRow } from '../settings/SettingsTappableRow'

interface Props extends EdgeAppSceneProps<'assetSettings'> {}

export function AssetSettingsScene(props: Props) {
  const { navigation } = props
  const theme = useTheme()
  const styles = getStyles(theme)

  const account = useSelector(state => state.core.account)
  const { currencyWallets } = account

  /**
   * Iterates through all the user's currency wallets, updates the enabled token
   * IDs to include any newly detected tokens, and then displays a toast message
   * indicating whether any detected tokens were enabled.
   */
  const handleRescanTokens = useHandler(async () => {
    let isTokensDetected = false
    for (const wallet of Object.values(currencyWallets)) {
      const detectedDisabledTokens = wallet.detectedTokenIds.filter(dt => !wallet.enabledTokenIds.includes(dt))

      if (detectedDisabledTokens.length > 0) {
        await wallet.changeEnabledTokenIds([...wallet.enabledTokenIds, ...detectedDisabledTokens])
        isTokensDetected = true
      }
    }

    if (isTokensDetected) {
      showToast(lstrings.settings_enable_detected_tokens_toast)
    } else {
      showToast(lstrings.settings_no_detected_tokens_toast)
    }
  })

  return (
    <SceneWrapper scroll>
      <SettingsTappableRow key="detectTokens" label={lstrings.settings_detect_tokens} onPress={handleRescanTokens}>
        <FontAwesomeIcon style={styles.icon} name="refresh" size={theme.rem(1.25)} color={theme.iconTappable} />
      </SettingsTappableRow>
      {CURRENCY_SETTINGS_KEYS.map(pluginId => {
        if (account.currencyConfig[pluginId] == null) return null
        const { currencyInfo } = account.currencyConfig[pluginId]
        const { displayName } = currencyInfo
        const onPress = () =>
          navigation.navigate('currencySettings', {
            currencyInfo
          })

        return (
          <SettingsTappableRow key={pluginId} label={displayName} onPress={onPress}>
            <CryptoIcon marginRem={[0.5, 0, 0.5, 0.5]} pluginId={pluginId} tokenId={null} sizeRem={1.25} />
          </SettingsTappableRow>
        )
      })}
    </SceneWrapper>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  icon: {
    marginLeft: theme.rem(0.6)
  }
}))
