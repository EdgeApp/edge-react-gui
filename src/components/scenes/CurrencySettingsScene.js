// @flow

import { asArray, asBoolean, asMaybe, asObject, asOptional, asString } from 'cleaners'
import type { EdgeAccount, EdgeDenomination } from 'edge-core-js'
import * as React from 'react'
import { ScrollView, Text, TouchableOpacity } from 'react-native'

import { setDenominationKeyRequest } from '../../actions/SettingsActions.js'
import s from '../../locales/strings.js'
import { getDisplayDenomination } from '../../selectors/DenominationSelectors.js'
import { useEffect, useState } from '../../types/reactHooks.js'
import { connect } from '../../types/reactRedux.js'
import { type RouteProp } from '../../types/routerTypes.js'
import { SceneWrapper } from '../common/SceneWrapper.js'
import { TextInputModal } from '../modals/TextInputModal.js'
import { Airship } from '../services/AirshipInstance.js'
import { type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'
import { SettingsHeaderRow } from '../themed/SettingsHeaderRow.js'
import { SettingsRadioRow } from '../themed/SettingsRadioRow.js'
import { SettingsSwitchRow } from '../themed/SettingsSwitchRow.js'
import { SettingsTappableRow } from '../themed/SettingsTappableRow.js'

type OwnProps = {
  route: RouteProp<'currencySettings'>
}
type StateProps = {
  account: EdgeAccount,
  selectedDenominationMultiplier: string
}
type DispatchProps = {
  selectDenomination: (pluginId: string, currencyCode: string, denomination: EdgeDenomination) => Promise<void>
}
type Props = StateProps & DispatchProps & ThemeProps & OwnProps

const asElectrumDefaults = asObject({
  electrumServers: asArray(asString)
})

const asElectrumSettings = asObject({
  disableFetchingServers: asOptional(asBoolean),
  electrumServers: asOptional(asArray(asString))
})

export function CurrencySettingsComponent(props: Props) {
  const { account, selectDenomination, selectedDenominationMultiplier, theme, route } = props
  const { currencyInfo } = route.params
  const { currencyCode, defaultSettings, denominations, pluginId } = currencyInfo
  const currencyConfig = account.currencyConfig[pluginId]
  const styles = getStyles(theme)

  // Follow the on-disk currency settings:
  const [userSettings = {}, setUserSettings] = useState(currencyConfig.userSettings)
  useEffect(() => currencyConfig.watch('userSettings', setUserSettings), [currencyConfig])

  // Are the electrum servers enabled? What are they?
  const defaults = asMaybe(asElectrumDefaults)(defaultSettings)
  const settings = asMaybe(asElectrumSettings)(userSettings)

  function renderCustomNodes(defaults: $Call<typeof asElectrumDefaults>, settings: $Call<typeof asElectrumSettings>) {
    const { electrumServers: defaultServers } = defaults
    const { disableFetchingServers = false, electrumServers = [] } = settings

    async function handleToggleNodes(): Promise<void> {
      await currencyConfig.changeUserSettings({
        ...currencyConfig.userSettings,
        disableFetchingServers: !disableFetchingServers,
        electrumServers: electrumServers.length > 0 ? electrumServers : defaultServers
      })
    }

    async function handleDeleteNode(i: number): Promise<void> {
      const list = [...electrumServers]
      list.splice(i, 1)

      await currencyConfig.changeUserSettings({
        ...currencyConfig.userSettings,
        electrumServers: list
      })
    }

    function handleEditNode(i?: number): void {
      async function handleSubmit(text: string) {
        const list = [...electrumServers]
        if (i == null) list.push(text)
        else list[i] = text

        await currencyConfig.changeUserSettings({
          ...currencyConfig.userSettings,
          electrumServers: list
        })
        return true
      }

      Airship.show(bridge => (
        <TextInputModal
          autoCorrect={false}
          bridge={bridge}
          initialValue={i == null ? '' : electrumServers[i]}
          inputLabel={s.strings.settings_custom_node_url}
          title={s.strings.settings_edit_custom_node}
          onSubmit={handleSubmit}
        />
      ))
    }

    return (
      <>
        <SettingsHeaderRow label={s.strings.settings_custom_nodes_title} />
        <SettingsSwitchRow label={s.strings.settings_enable_custom_nodes} value={disableFetchingServers} onPress={handleToggleNodes} />
        {!disableFetchingServers ? null : (
          <>
            {electrumServers.map((server, i) => (
              <SettingsTappableRow key={`row${i}`} action="delete" onPress={() => handleDeleteNode(i)}>
                <TouchableOpacity onPress={() => handleEditNode(i)} style={styles.labelContainer}>
                  <Text style={styles.labelText}>{server}</Text>
                </TouchableOpacity>
              </SettingsTappableRow>
            ))}
            <SettingsTappableRow action="add" label={s.strings.settings_add_custom_node} onPress={handleEditNode} />
          </>
        )}
      </>
    )
  }

  return (
    <SceneWrapper background="theme" hasTabs={false}>
      <ScrollView>
        <SettingsHeaderRow label={s.strings.settings_denominations_title} />
        {denominations.map(denomination => {
          const key = denomination.multiplier
          const isSelected = key === selectedDenominationMultiplier
          return (
            <SettingsRadioRow key={denomination.multiplier} value={isSelected} onPress={() => selectDenomination(pluginId, currencyCode, denomination)}>
              <Text style={styles.labelText}>
                <Text style={styles.symbolText}>{denomination.symbol}</Text>
                {' - ' + denomination.name}
              </Text>
            </SettingsRadioRow>
          )
        })}
        {defaults == null || settings == null ? null : renderCustomNodes(defaults, settings)}
      </ScrollView>
    </SceneWrapper>
  )
}

const getStyles = cacheStyles(theme => ({
  // We use a hack to make the text tappable separately from the switch.
  labelContainer: {
    flexGrow: 10,
    flexShrink: 1,
    // Stretch outward to cover the row:
    margin: -theme.rem(0.5),
    padding: theme.rem(0.5)
  },
  labelText: {
    color: theme.primaryText,
    flexShrink: 1,
    fontFamily: theme.fontFaceDefault,
    fontSize: theme.rem(1),
    paddingHorizontal: theme.rem(0.5),
    textAlign: 'left'
  },
  symbolText: {
    fontFamily: theme.fontFaceSymbols
  }
}))

export const CurrencySettingsScene = connect<StateProps, DispatchProps, OwnProps>(
  (state, { route: { params } }) => ({
    account: state.core.account,
    selectedDenominationMultiplier: getDisplayDenomination(state, params.currencyInfo.pluginId, params.currencyInfo.currencyCode).multiplier
  }),
  dispatch => ({
    async selectDenomination(pluginId, currencyCode, denomination) {
      await dispatch(setDenominationKeyRequest(pluginId, currencyCode, denomination))
    }
  })
)(withTheme(CurrencySettingsComponent))
