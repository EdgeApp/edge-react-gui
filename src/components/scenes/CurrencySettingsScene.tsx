import * as React from 'react'
import { ScrollView, Text } from 'react-native'

import { setDenominationKeyRequest } from '../../actions/SettingsActions'
import { lstrings } from '../../locales/strings'
import { getDisplayDenomination } from '../../selectors/DenominationSelectors'
import { useDispatch, useSelector } from '../../types/reactRedux'
import { EdgeSceneProps } from '../../types/routerTypes'
import { SceneWrapper } from '../common/SceneWrapper'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { SettingsHeaderRow } from '../settings/SettingsHeaderRow'
import { SettingsRadioRow } from '../settings/SettingsRadioRow'
import { MaybeBlockbookSetting, MaybeCustomServersSetting, MaybeElectrumSetting } from '../themed/MaybeCustomServersSetting'
import { MaybeMoneroUserSettings } from '../themed/MaybeMoneroUserSettings'

interface Props extends EdgeSceneProps<'currencySettings'> {}

export function CurrencySettingsScene(props: Props) {
  const { route } = props
  const { currencyInfo } = route.params
  const { currencyCode, denominations, pluginId } = currencyInfo
  const theme = useTheme()
  const styles = getStyles(theme)
  const dispatch = useDispatch()

  const selectedDenominationMultiplier = useSelector(state => getDisplayDenomination(state, pluginId, currencyCode).multiplier)
  const account = useSelector(state => state.core.account)
  const currencyConfig = account.currencyConfig[pluginId]

  function renderDenominations() {
    return (
      <>
        <SettingsHeaderRow label={lstrings.settings_denominations_title} />
        {denominations.map(denomination => {
          const key = denomination.multiplier
          const isSelected = key === selectedDenominationMultiplier
          const handlePress = async () => {
            await dispatch(setDenominationKeyRequest(pluginId, currencyCode, denomination))
          }

          return (
            <SettingsRadioRow key={key} value={isSelected} onPress={handlePress}>
              <Text style={styles.labelText}>
                <Text style={styles.symbolText}>{denomination.symbol}</Text>
                {' - ' + denomination.name}
              </Text>
            </SettingsRadioRow>
          )
        })}
      </>
    )
  }

  return (
    <SceneWrapper background="theme" hasTabs={false}>
      <ScrollView>
        {denominations.length > 1 ? renderDenominations() : null}
        <MaybeBlockbookSetting currencyConfig={currencyConfig} />
        <MaybeCustomServersSetting currencyConfig={currencyConfig} />
        <MaybeElectrumSetting currencyConfig={currencyConfig} />
        <MaybeMoneroUserSettings currencyConfig={currencyConfig} />
      </ScrollView>
    </SceneWrapper>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  labelText: {
    color: theme.primaryText,
    flexShrink: 1,
    fontFamily: theme.fontFaceDefault,
    fontSize: theme.rem(1),
    paddingHorizontal: theme.rem(0.5),
    textAlign: 'left'
  },
  symbolText: {
    fontFamily: theme.fontFaceSymbols
  }
}))
