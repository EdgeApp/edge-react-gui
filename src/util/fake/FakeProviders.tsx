import * as React from 'react'
import { Metrics, SafeAreaProvider } from 'react-native-safe-area-context'
import { Provider } from 'react-redux'
import { applyMiddleware, createStore } from 'redux'
import thunk from 'redux-thunk'

import { rootReducer, RootState } from '../../reducers/RootReducer'

type DeepPartial<T> = T extends object
  ? {
      [P in keyof T]?: DeepPartial<T[P]>
    }
  : T

export type FakeState = DeepPartial<RootState>

interface Props {
  children: React.ReactNode
  initialState?: FakeState
}

export function FakeProviders(props: Props) {
  const { children, initialState = {} } = props

  const store = React.useMemo(() => createStore(rootReducer, initialState as any, applyMiddleware(thunk)), [initialState])
  return (
    <SafeAreaProvider initialMetrics={initialMetrics}>
      <Provider store={store}>{children}</Provider>
    </SafeAreaProvider>
  )
}

// These match what our old react-native-safe-area-view library returned:
const initialMetrics: Metrics = {
  frame: { height: 1334, width: 750, x: 0, y: 0 },
  insets: { bottom: 0, left: 0, right: 0, top: 20 }
}
