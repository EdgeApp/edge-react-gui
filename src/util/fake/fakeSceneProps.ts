import { AppParamList, EdgeSceneProps, NavigationProp } from '../../types/routerTypes'

export const fakeNavigation: NavigationProp<any> = {
  addListener() {
    return () => undefined
  },
  removeListener() {},
  isFocused() {
    return true
  },

  navigate() {},
  push() {},
  replace() {},
  setParams(params) {},

  goBack() {},
  pop() {},
  popToTop() {},

  dispatch() {},
  reset() {},
  canGoBack() {
    return true
  },
  getId() {
    return undefined
  },
  getState() {
    return {
      key: 'foo',
      index: 0,
      routes: [],
      routeNames: [],
      type: 'bar',
      stale: false
    }
  },
  getParent() {
    throw new Error('not implemented')
  },
  setOptions() {}
}

export function fakeSceneProps<Name extends keyof AppParamList>(name: Name, params: AppParamList[Name]): EdgeSceneProps<Name> {
  return {
    navigation: fakeNavigation,
    route: {
      name: name,
      key: name + '-0',
      params
    } as any
  }
}
