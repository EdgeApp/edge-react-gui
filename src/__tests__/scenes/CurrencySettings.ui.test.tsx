import { describe, expect, it } from '@jest/globals'
import * as React from 'react'
import TestRenderer from 'react-test-renderer'

import { CurrencySettingsScene } from '../../components/scenes/CurrencySettingsScene'
import { FakeProviders, FakeState } from '../../util/fake/FakeProviders'
import { fakeSceneProps } from '../../util/fake/fakeSceneProps'

describe('CurrencySettings', () => {
  it('should render', () => {
    const currencyInfo: any = {
      currencyCode: 'BTG',
      defaultSettings: {},
      denominations: [
        { name: 'BTG', multiplier: '100000000', symbol: '₿' },
        { name: 'mBTG', multiplier: '100000', symbol: 'm₿' },
        { name: 'bits', multiplier: '100', symbol: 'ƀ' }
      ],
      pluginId: 'bitcoin-gold'
    }

    const state: FakeState = {
      core: {
        account: {
          currencyConfig: {
            'bitcoin-gold': {
              currencyInfo,
              watch: () => () => {}
            }
          }
        }
      }
    }

    const renderer = TestRenderer.create(
      <FakeProviders initialState={state}>
        <CurrencySettingsScene
          {...fakeSceneProps('currencySettings', {
            currencyInfo
          })}
        />
      </FakeProviders>
    )

    expect(renderer.toJSON()).toMatchSnapshot()
    renderer.unmount()
  })
})
