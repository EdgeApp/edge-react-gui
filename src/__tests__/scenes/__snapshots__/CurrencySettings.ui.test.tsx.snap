// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencySettings should render 1`] = `
<BVLinearGradient
  colors={
    [
      4278585612,
      4279971643,
    ]
  }
  endPoint={
    {
      "x": 1,
      "y": 0,
    }
  }
  locations={null}
  startPoint={
    {
      "x": 0,
      "y": 0,
    }
  }
  style={
    {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "stretch",
          "flexDirection": "column",
          "justifyContent": "flex-start",
          "position": "absolute",
        },
        {
          "bottom": 0,
          "left": 0,
          "padding": 0,
          "right": 0,
          "top": 64,
        },
      ]
    }
  >
    <RCTScrollView>
      <View>
        <BVLinearGradient
          colors={
            [
              4279377189,
              4279377189,
            ]
          }
          endPoint={
            {
              "x": 1,
              "y": 1,
            }
          }
          locations={null}
          startPoint={
            {
              "x": 0,
              "y": 0,
            }
          }
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "flex-start",
              "minHeight": 62,
              "padding": 22,
            }
          }
        >
          <Text
            style={
              {
                "color": "#FFFFFF",
                "flexGrow": 1,
                "flexShrink": 1,
                "fontFamily": "Quicksand-Medium",
                "fontSize": 22,
                "textAlign": "left",
              }
            }
          >
            Denominations
          </Text>
        </BVLinearGradient>
        <View
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(255, 255, 255, 0)",
              "flexDirection": "row",
              "marginBottom": 1,
              "minHeight": 67,
              "padding": 11,
            }
          }
        >
          <Text
            style={
              {
                "color": "#FFFFFF",
                "flexShrink": 1,
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "paddingHorizontal": 11,
                "textAlign": "left",
              }
            }
          >
            <Text
              style={
                {
                  "fontFamily": "Quicksand-Regular",
                }
              }
            >
              ₿
            </Text>
             - BTG
          </Text>
          <Text
            style={
              {
                "color": "#FFFFFF",
                "flexGrow": 1,
                "flexShrink": 1,
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "paddingHorizontal": 11,
                "textAlign": "left",
              }
            }
          />
          <View>
            <View
              pending={false}
              style={
                [
                  {
                    "aspectRatio": 1,
                    "position": "absolute",
                    "width": "100%",
                    "zIndex": 1,
                  },
                  {
                    "opacity": 0,
                  },
                ]
              }
            >
              <ActivityIndicator
                color="#00f1a2"
                style={
                  {
                    "height": 34,
                    "marginHorizontal": 11,
                  }
                }
              />
            </View>
            <View
              pending={false}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <Text
                accessibilityHint="On undefined"
                accessibilityRole="radio"
                allowFontScaling={false}
                style={
                  [
                    {
                      "color": undefined,
                      "fontSize": 12,
                    },
                    {
                      "color": "#00f1a2",
                      "fontSize": 28,
                      "marginHorizontal": 11,
                    },
                    {
                      "fontFamily": "Ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
            </View>
          </View>
        </View>
        <View
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(255, 255, 255, 0)",
              "flexDirection": "row",
              "marginBottom": 1,
              "minHeight": 67,
              "padding": 11,
            }
          }
        >
          <Text
            style={
              {
                "color": "#FFFFFF",
                "flexShrink": 1,
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "paddingHorizontal": 11,
                "textAlign": "left",
              }
            }
          >
            <Text
              style={
                {
                  "fontFamily": "Quicksand-Regular",
                }
              }
            >
              m₿
            </Text>
             - mBTG
          </Text>
          <Text
            style={
              {
                "color": "#FFFFFF",
                "flexGrow": 1,
                "flexShrink": 1,
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "paddingHorizontal": 11,
                "textAlign": "left",
              }
            }
          />
          <View>
            <View
              pending={false}
              style={
                [
                  {
                    "aspectRatio": 1,
                    "position": "absolute",
                    "width": "100%",
                    "zIndex": 1,
                  },
                  {
                    "opacity": 0,
                  },
                ]
              }
            >
              <ActivityIndicator
                color="#00f1a2"
                style={
                  {
                    "height": 34,
                    "marginHorizontal": 11,
                  }
                }
              />
            </View>
            <View
              pending={false}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <Text
                accessibilityHint="Off undefined"
                accessibilityRole="radio"
                allowFontScaling={false}
                style={
                  [
                    {
                      "color": undefined,
                      "fontSize": 12,
                    },
                    {
                      "color": "#00f1a2",
                      "fontSize": 28,
                      "marginHorizontal": 11,
                    },
                    {
                      "fontFamily": "Ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
            </View>
          </View>
        </View>
        <View
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(255, 255, 255, 0)",
              "flexDirection": "row",
              "marginBottom": 1,
              "minHeight": 67,
              "padding": 11,
            }
          }
        >
          <Text
            style={
              {
                "color": "#FFFFFF",
                "flexShrink": 1,
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "paddingHorizontal": 11,
                "textAlign": "left",
              }
            }
          >
            <Text
              style={
                {
                  "fontFamily": "Quicksand-Regular",
                }
              }
            >
              ƀ
            </Text>
             - bits
          </Text>
          <Text
            style={
              {
                "color": "#FFFFFF",
                "flexGrow": 1,
                "flexShrink": 1,
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "paddingHorizontal": 11,
                "textAlign": "left",
              }
            }
          />
          <View>
            <View
              pending={false}
              style={
                [
                  {
                    "aspectRatio": 1,
                    "position": "absolute",
                    "width": "100%",
                    "zIndex": 1,
                  },
                  {
                    "opacity": 0,
                  },
                ]
              }
            >
              <ActivityIndicator
                color="#00f1a2"
                style={
                  {
                    "height": 34,
                    "marginHorizontal": 11,
                  }
                }
              />
            </View>
            <View
              pending={false}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <Text
                accessibilityHint="Off undefined"
                accessibilityRole="radio"
                allowFontScaling={false}
                style={
                  [
                    {
                      "color": undefined,
                      "fontSize": 12,
                    },
                    {
                      "color": "#00f1a2",
                      "fontSize": 28,
                      "marginHorizontal": 11,
                    },
                    {
                      "fontFamily": "Ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</BVLinearGradient>
`;
